test/src/unit-conversions.cpp

```diff
@@ -1703,6 +1703,65 @@ TEST_CASE("JSON to enum mapping")
     }
 }
 
+#ifdef JSON_HAS_CPP_17
+TEST_CASE("std::optional")
+{
+    SECTION("null")
+    {
+        json j_null;
+        std::optional<std::string> opt_null;
+
+        CHECK(json(opt_null) == j_null);
+        CHECK(std::optional<std::string>(j_null) == std::nullopt);
```

Comment:
Initialization of `optional<T>` from empty `json` should be considered as a `type_error`:
```suggestion
        CHECK_THROWS_AS(std::optional<std::string>(j_null), json::type_error&);
```
The reason is as follows. If `json` is implicitly convertible to `T`, then initialization of `optional<T>` from `json` almost necessarily goes through the "forwarding" constructor:
```
template <class U> optional(U&& value); // unrelated details omitted
```
This constructor provides the best parameter/argument match by value category, constness and a type. Its parameter binds directly to `json` in any form. The only "weakness" of the forwarding constructor is that it is a template. A conversion function that would be considered as a better candidate for overload resolution must be non-templated, but this is obviously unacceptable.

The forwarding constructor converts `json` to `T` (initializes the latter from the former) and stores the result, so the constructed `optional<T>` cannot be empty. Such a conversion from empty `json` to `T` is a `type error` (even if `T` itself is `optional<X>`, by induction).

The non-standard overload resolution performed by GCC and Clang when enabling C++17 does not change the outcome. The only way to achieve the "natural" behavior is to tune constructors of `std::optional<T>`, which is beyond the scope of the project.

Reply:
There is also the usual way to explicitly convert `json` to `optional<T>`. The following will work:
```
        CHECK(j_null.get<std::optional<std::string>>() == std::nullopt);
```

Reply:
Essentially, there is nothing new here. @dota17 proposed the same in #2213.
However, this approach (treat such a conversion as `type_error`) introduces some inconsistency:
```
std::optional<int> opt1 = json().get<std::optional<int>>(); // ok: std::nullopt
std::optional<int> opt2 = std::optional<int>(json());       // throws type_error
std::optional<int> opt3 = json();                           // throws type_error
```
Perhaps it makes sense to throw `type_error` in the first case too?
The "natural" conversion could be explicitly implemented by dedicated member function:
```
template <typename T> std::optional<T> get_optional() const;
```

Reply:
Moreover, as it turns out, this dedicated member function is the only meaningful thing `nlohmann::basic_json` can provide to support `std::optional`. Overloads of `get<std::optional<T>>` should probably be disabled to prevent confusion.

Reply:
I think the behavior is counter-intuitive. JSON's `null` is usually considered as "unset" value (in contrast to an empty value). Therefore, I would expect `null` to convert to `std::nullopt` of any optional type.